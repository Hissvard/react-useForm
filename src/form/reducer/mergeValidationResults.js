import { Map, List } from 'immutable'

const getError = (errors, key) => {
  const error = errors[key]

  if (typeof error === 'string') return error
  if (Array.isArray(error) && error.error) {
    return error.error
  }

  return ''
}

const updateItems = (list, errorObj) => {
  const errors = errorObj || {}

  return list.map((listItem, index) => {
    const error = Array.isArray(errors) ? errors[index] : errors
    return listItem.update('fields', Map(), fields => updateMap(fields, error))
  })
}

const updateMap = (fields, errorObj) => {
  const errors = errorObj || {}

  return fields.entrySeq().reduce((acc, [key, field]) => {
    const next = field.get('items') ?
      field.update('items', List(), items => updateItems(items, errors[key])) :
      field


    const error = getError(errors, key)

    return acc
      .mergeIn([key], next)
      .setIn([key, 'current', 'error'], Boolean(error))
      .setIn([key, 'current', 'helperText'], error)

  }, Map())

}

export const mergeValidationResults = (state, errors = {}) => {
  return state.update('fields', Map(), fields => updateMap(fields, errors))
}
