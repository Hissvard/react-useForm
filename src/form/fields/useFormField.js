import { actions } from '../reducer/fieldReducer'
/**
 * @param  fieldArgs { import("../useForm").IFormFieldArgs}
 * @returns {import("../useForm").IFormField}
 */
export const useFormField = (state, dispatch, fieldArgs = {}) => {
  const requiredMessage = fieldArgs.requiredMessage || 'Required'

  const validate = (value, touched = false) => {
    let result

    if (value !== '' && value !== undefined && fieldArgs.validate) {
      result = fieldArgs.validate(value, fieldArgs.name, state.get('getAllValues'))
    }

    if (!Boolean(result) && !fieldArgs.optional && touched && value === '') {
      result = requiredMessage
    }

    if (touched || (!Boolean(result) && state.getIn(['current', 'helperText']))) {
      dispatch(actions.validationResult(fieldArgs.name, Boolean(result), result))
    }
    return result
  }

  const setValue = v => {
    dispatch(actions.updateValue(fieldArgs.name, v))
    validate(v, state.getIn(['current', 'touched']))
  }

  const onChange = event => {
    const value = fieldArgs.valueFromChange ? fieldArgs.valueFromChange(event) : event.target.value
    const coercedValue = fieldArgs.normalize ? fieldArgs.normalize(value) : value
    setValue(coercedValue)
    validate(coercedValue, state.getIn(['current', 'touched']))
  }

  const onBlur = () => {
    dispatch(actions.touched(fieldArgs.name))
    validate(state.getIn(['current', 'value']), true)
  }

  return {
    props: {
      error: state.getIn(['current', 'error']),
      helperText: state.getIn(['current', 'helperText']),
      label: state.getIn(['initial', 'label']),
      value: state.getIn(['current', 'value']),
      // handlers
      onBlur,
      onChange,
    },
    meta: {
      touched: state.getIn(['current', 'touched']),
      pristine: state.getIn(['current', 'pristine']),
    },
    setValue,
  }
}
