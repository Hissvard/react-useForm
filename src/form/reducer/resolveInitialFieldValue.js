import { defaultTextValue } from '../fields/useTextField'
import { defaultSelectValue } from '../fields/useSelectField'
import { defaultNumberValue } from '../fields/useNumberField'
import { defaultBooleanValue } from '../fields/useBooleanField'
import { defaultListValue } from '../fields/useListField'
import { defaultObjectValue } from '../fields/useObjectField'

const defaultValues = {
  text: defaultTextValue,
  select: defaultSelectValue,
  number: defaultNumberValue,
  boolean: defaultBooleanValue,
  list: defaultListValue,
  object: defaultObjectValue,
}

export const resolveInitialFieldValue = (fieldValue, initialValue, type = 'text') => {
  if (initialValue !== undefined) {
    return initialValue
  }
  if (fieldValue !== undefined) {
    return fieldValue
  }

  const defaultValue = defaultValues[type]
  return defaultValue !== undefined ? defaultValue : defaultTextValue
}
