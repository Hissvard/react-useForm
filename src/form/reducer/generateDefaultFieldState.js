import { fromJS } from 'immutable'
import { resolveInitialFieldValue } from './resolveInitialFieldValue'
import { resolveLabel } from './resolveLabel'

/**
 *
 * @param field {import('../useForm').IFormFieldArgs}
 * @param {*} initialValues
 * @param options {import('../useForm').IFormOptions}
 */
export const generateDefaultFieldState = (field, initialValues, options) => {
  if (!field.name) throw new Error('name is required on field')

  const label = resolveLabel(field, options)
  const value = resolveInitialFieldValue(field.value, initialValues.getIn(field.name.split('.')), field.type)

  return fromJS({
    initial: {
      type: field.type || 'text',
      optional: field.optional || false,
      label,
    },
    current: {
      helperText: field.helperText || '',
      error: false,
      pristine: true,
      touched: false,
    },
  }).setIn(['initial', 'field'], field)
    .setIn(['initial', 'value'], value)
    .setIn(['current', 'value'], value)
}
