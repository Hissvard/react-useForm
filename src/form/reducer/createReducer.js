import { Map } from 'immutable'
import { useReducer, useMemo } from 'react'
import { fieldReducer, actions } from './fieldReducer'
import { getInitialState } from './getInitialState'

export const createReducer = ({ fields, options = {}, initialValues = Map(), formTools }) => {
  const calcInitialState = () => {
    return getInitialState(fields, initialValues, options, formTools)
  }
  const [state, dispatch] = useReducer(fieldReducer, Map(), () => {
    return calcInitialState()
  })

  useMemo(() => {
    setTimeout(() => {
      dispatch(actions.reset(calcInitialState()))
    })
  }, [initialValues.hashCode()])

  return [state, dispatch]
}
