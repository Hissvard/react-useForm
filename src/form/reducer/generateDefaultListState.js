import { fromJS, Map } from 'immutable'
import { resolveInitialFieldValue } from './resolveInitialFieldValue'
import { resolveLabel } from './resolveLabel'
import { getFieldState } from './getFieldState'

/**
 *
 * @param field {import("..").IFormFieldArgs}
 * @param {*} initialValues
 * @param options {import("..").IFormOptions}
 */
export const generateDefaultListState = (field, initialValues, options = {}, parentPath = '') => {
  // const applicablePath = parentPath ? `${parentPath}.${field.name}` : field.name
  const label = resolveLabel(field, options)
  const listValues = resolveInitialFieldValue(field.value, initialValues.getIn(field.name.split('.')), field.type)
  const items = getCurrentValues(field.fields, listValues, options, field.name, parentPath)

  return fromJS({
    initial: {
      type: field.type,
      optional: field.optional || false,
      label,
    },
    current: {
      helperText: field.helperText || '',
      error: false,
      pristine: true,
      touched: false,
    },
    items,
  }).setIn(['initial', 'field'], field)
    .setIn(['initial', 'options'], options)
}

export const getCurrentValues = (fields = [], initialListValues, options, fieldPath, parentPath) => {
  return initialListValues.map((item, index) => {
    return getFields(fields, item, options, fieldPath, index, parentPath)
  })
}

export const getFields = (fields = [], value = Map(), options, fieldPath, index = 0, parentPath = '') => {

  return Map([['fields', fields.reduce((acc, field) => {
    let pathName = `${fieldPath}.items.${index}.fields`
    if (parentPath) pathName = `${parentPath}.${pathName}`
    return acc.set(field.name, getFieldState(field, value, options, pathName))
      .setIn([field.name, 'initial', 'field'], {
        ...field,
        name: `${pathName}.${field.name}`,
      })
  }, Map())]])
}
