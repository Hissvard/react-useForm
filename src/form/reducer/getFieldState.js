import { generateDefaultListState } from './generateDefaultListState'
import { generateDefaultFieldState } from './generateDefaultFieldState'

export const getFieldState = (field, initialValues, options, parentPath = '') => {
  if (field.type === 'list') {
    return generateDefaultListState(field, initialValues, options, parentPath)
  }
  return generateDefaultFieldState(field, initialValues, options)
}
