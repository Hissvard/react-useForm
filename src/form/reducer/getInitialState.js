import { fromJS, Map, List } from 'immutable'
import { fieldsKey, listenersKey } from './fieldReducer'
import { generateDefaultListState } from './generateDefaultListState'
import { generateDefaultFieldState } from './generateDefaultFieldState'

/**
 * @param fields {Array<import("..").IFormFieldArgs>}
 */
export const getInitialState = (fields, initialValues = Map(), options = {}) => {
  const fieldMap = fields.reduce((acc, field) => {
    const result = acc.set(field.name, getFieldState(field, initialValues, options))

    return result
  }, Map())

  const listeners = getListeners(options.listeners)
  return fromJS({ [fieldsKey]: fieldMap }).merge(listeners)
}

const getFieldState = (field, initialValues, options) => {
  if (field.type === 'list') {
    return generateDefaultListState(field, initialValues, options)
  }
  return generateDefaultFieldState(field, initialValues, options)
}

const getListeners = (listeners = {}) => {
  return Object.entries(listeners).reduce((acc, [fieldName, value]) => {
    return acc.updateIn([listenersKey, fieldName], List(), current => {
      return Array.isArray(value) ? current.concat(value) : current.push(value)
    })
  }, Map())
}
