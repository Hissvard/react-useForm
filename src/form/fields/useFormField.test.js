import { fromJS } from 'immutable'
import { useFormField } from './useFormField'
import { actionTypes } from '../reducer/fieldReducer'

jest.mock('react')

const dispatch = jest.fn()
let value, initialArgs, state
beforeEach(() => {
  value = 'the value'
  initialArgs = {
    name: 'fieldName',
    optional: false,
    error: 'on',
    helperText: 'the help',
    label: 'the label',
  }

  state = fromJS({
    initial: {
      label: initialArgs.label,
      type: 'text',
      value,
      optional: false,
    },
    current: {
      value,
      error: false,
      pristine: true,
      touched: false,
      helperText: initialArgs.helperText,
    },
    getAllValues: 'getAllValuesFunc',
  })
})

test('returns all props needed', () => {
  expect(useFormField(state, dispatch, initialArgs)).toMatchSnapshot()
})

describe('onChange', () => {
  let event
  beforeEach(() => {
    event = {
      target: {
        value: 'updated',
      },
    }
  })

  test('dispatches change action', () => {
    const { props: { onChange } } = useFormField(state, dispatch, initialArgs)
    onChange(event)

    expect(dispatch.mock.calls[0]).toMatchSnapshot()
  })

  test('uses valueFromChange to coerce value if provided', () => {
    const valueFromChange = jest.fn()
    valueFromChange.mockReturnValue('updatedFromTarget')
    const { props: { onChange } } = useFormField(state, dispatch, { ...initialArgs, valueFromChange })
    onChange(event)

    expect(dispatch.mock.calls[0]).toMatchSnapshot()
    expect(valueFromChange).toHaveBeenCalledWith(event)
  })

  test('coerces value using normalize', () => {
    const normalize = jest.fn()
    normalize.mockReturnValue('normalized value')
    const { props: { onChange } } = useFormField(state, dispatch, { ...initialArgs, normalize })
    onChange(event)

    expect(dispatch.mock.calls[0]).toMatchSnapshot()
    expect(normalize).toHaveBeenCalledWith(event.target.value)
  })

  test('validates on change', () => {
    initialArgs.validate = jest.fn()
    const { props: { onChange } } = useFormField(state, dispatch, initialArgs)
    onChange(event)

    expect(initialArgs.validate.mock.calls[0]).toMatchSnapshot()
  })

  test('does not dispatch validation result until "touched"', () => {
    initialArgs.validate = jest.fn()
    initialArgs.validate.mockReturnValue('new error')
    state = state.setIn(['current', 'helperText'], 'error exists')

    const { props: { onChange } } = useFormField(state, dispatch, initialArgs)
    onChange(event)

    expect(dispatch.mock.calls).toMatchSnapshot()
  })

  test('clears result event if not touched yet', () => {
    initialArgs.validate = jest.fn()
    initialArgs.validate.mockReturnValue('')
    state = state.setIn(['current', 'helperText'], 'error exists')

    const { props: { onChange } } = useFormField(state, dispatch, initialArgs)
    onChange(event)

    expect(dispatch.mock.calls).toMatchSnapshot()
    expect(dispatch).toHaveBeenCalledTimes(2)
  })

  test('dispatches validation result if "touched"', () => {
    state = state.setIn(['current', 'touched'], true)
    initialArgs.validate = jest.fn()
    const { props: { onChange } } = useFormField(state, dispatch, initialArgs)
    onChange(event)

    expect(dispatch.mock.calls).toMatchSnapshot()
  })
})

describe('validate', () => {
  const validate = jest.fn()
  let event
  beforeEach(() => {
    initialArgs = { ...initialArgs, validate }
    event = {
      target: {
        value: '1',
      },
    }
    state = fromJS({
      current: {
        value,
        error: false,
        pristine: false,
        touched: true,
        helperText: initialArgs.helperText,
      },
      getAllValues: 'getAllValuesFunc',
    })
  })

  test('calls validate on change', () => {
    const { props: { onChange } } = useFormField(state, dispatch, initialArgs)
    onChange(event)

    expect(validate.mock.calls[0]).toMatchSnapshot()
  })

  test('calls validate on number change with 0 (falsy)', () => {
    initialArgs.normalize = v => Number(v)
    event.target.value = '0'
    const { props: { onChange } } = useFormField(state, dispatch, initialArgs)
    onChange(event)

    expect(validate.mock.calls[0]).toMatchSnapshot()
  })

  test('does not set error if not "touched"', () => {
    state = fromJS({
      current: {
        value,
        error: false,
        pristine: false,
        touched: false,
        helperText: initialArgs.helperText,
      },
    })
    validate.mockReturnValue('Error here')

    const { props: { onChange } } = useFormField(state, dispatch, initialArgs)
    onChange(event)

    expect(dispatch).not.toHaveBeenCalledWith({
      type: actionTypes.validationResult,
      payload: {
        error: true,
        helperText: expect.any(String),
      },
    })

  })

  test('sets validation result if "touched"', () => {
    validate.mockReturnValue('Error here')

    const { props: { onChange } } = useFormField(state, dispatch, initialArgs)
    onChange(event)

    expect(dispatch).toHaveBeenCalledWith({
      type: actionTypes.validationResult,
      fieldName: 'fieldName',
      payload: {
        error: true,
        helperText: expect.any(String),
      },
    })

  })

  test('validates onBlur', () => {
    validate.mockReturnValue('Error here')

    const { props: { onBlur } } = useFormField(state, dispatch, initialArgs)
    onBlur()

    expect(validate.mock.calls[0]).toMatchSnapshot()
  })

  test('sets required if empty on blur', () => {
    state = fromJS({
      current: {
        value: '',
        error: false,
        pristine: false,
        touched: true,
        helperText: initialArgs.helperText,
      },
    })

    const { props: { onBlur } } = useFormField(state, dispatch, initialArgs)
    onBlur()

    expect(dispatch).toHaveBeenCalledWith({
      type: actionTypes.validationResult,
      fieldName: 'fieldName',
      payload: {
        error: true,
        helperText: expect.any(String),
      },
    })
  })

  test('validate prepares name for reading', () => {
    initialArgs = { ...initialArgs, name: 'colors.items.1.fields.shades.items.0.fields.shade', validate }
    event = {
      target: {
        value: '1',
      },
    }
    state = fromJS({
      current: {
        value,
        error: false,
        pristine: false,
        touched: true,
        helperText: initialArgs.helperText,
      },
      getAllValues: 'getAllValuesFunc',
    })

    initialArgs.validate = jest.fn()
    const { props: { onChange } } = useFormField(state, dispatch, initialArgs)
    onChange(event)

    expect(initialArgs.validate.mock.calls[0]).toMatchSnapshot()
  })
})

test('sets touched onBlur', () => {
  const { props: { onBlur } } = useFormField(state, dispatch, initialArgs)
  onBlur()

  expect(dispatch).toHaveBeenCalledWith({
    fieldName: 'fieldName',
    type: actionTypes.touched,
  })
})

test('setValue sets value externally', () => {
  const { setValue } = useFormField(state, dispatch, initialArgs)

  setValue('i am value')
  expect(dispatch).toHaveBeenCalledWith({
    type: actionTypes.updateValue,
    fieldName: 'fieldName',
    payload: 'i am value',
  })
})

