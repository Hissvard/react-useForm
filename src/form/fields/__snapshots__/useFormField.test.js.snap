// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onChange clears result event if not touched yet 1`] = `
Array [
  Array [
    Object {
      "fieldName": "fieldName",
      "payload": "updated",
      "type": "updateValue",
    },
  ],
  Array [
    Object {
      "fieldName": "fieldName",
      "payload": Object {
        "error": false,
        "helperText": "",
      },
      "type": "validationResult",
    },
  ],
]
`;

exports[`onChange coerces value using normalize 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": "normalized value",
    "type": "updateValue",
  },
]
`;

exports[`onChange dispatches change action 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": "updated",
    "type": "updateValue",
  },
]
`;

exports[`onChange dispatches validation result if "touched" 1`] = `
Array [
  Array [
    Object {
      "fieldName": "fieldName",
      "payload": "updated",
      "type": "updateValue",
    },
  ],
  Array [
    Object {
      "fieldName": "fieldName",
      "payload": Object {
        "error": false,
        "helperText": undefined,
      },
      "type": "validationResult",
    },
  ],
]
`;

exports[`onChange does not dispatch validation result until "touched" 1`] = `
Array [
  Array [
    Object {
      "fieldName": "fieldName",
      "payload": "updated",
      "type": "updateValue",
    },
  ],
]
`;

exports[`onChange uses valueFromChange to coerce value if provided 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": "updatedFromTarget",
    "type": "updateValue",
  },
]
`;

exports[`onChange validates on change 1`] = `
Array [
  "updated",
  "fieldName",
  "getAllValuesFunc",
]
`;

exports[`returns all props needed 1`] = `
Object {
  "meta": Object {
    "pristine": true,
    "touched": false,
  },
  "props": Object {
    "error": false,
    "helperText": "the help",
    "label": "the label",
    "onBlur": [Function],
    "onChange": [Function],
    "value": "the value",
  },
  "setValue": [Function],
}
`;

exports[`validate calls validate on change 1`] = `
Array [
  "1",
  "fieldName",
  "getAllValuesFunc",
]
`;

exports[`validate calls validate on number change with 0 (falsy) 1`] = `
Array [
  0,
  "fieldName",
  "getAllValuesFunc",
]
`;

exports[`validate validates onBlur 1`] = `
Array [
  "the value",
  "fieldName",
  "getAllValuesFunc",
]
`;
